### PR TITLE
Features for downstream

### PR DESCRIPTION
**Pull request by jdolence**
## PR Summary

Two main improvements:

1. Adding a new overload to `DataCollection::Add` that allows new `MeshBlockData` or `MeshData` objects to be created from a base object, but including only specific named variables.
2. Add a slightly more general `SumData` function in `Update`

Also changed a lot of the API to `T *` instead of `std::shared_ptr<T> &`.  `const T *` would have been better, but the caching we do for packing prevents the usage of `const`.

## PR Checklist

- [x] Code passes cpplint
- [x] New features are documented.
- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Code is formatted
- [x] Changes are summarized in CHANGELOG.md

### CODE REVIEW

File: src/interface/mesh_data.hpp
Comment (JoshuaSBrown) [inline src/interface/mesh_data.hpp:82]: 👍🏻

File: src/interface/mesh_data.hpp
Comment (JoshuaSBrown) [inline src/interface/mesh_data.hpp:89]: 👍🏻

File: src/interface/meshblock_data.hpp
Comment (JoshuaSBrown) [inline src/interface/meshblock_data.hpp:70]: 👍🏻

File: src/interface/meshblock_data.hpp
Comment (JoshuaSBrown) [inline src/interface/meshblock_data.hpp]: label can be made const reference.

File: src/interface/meshblock_data.hpp
Comment (JoshuaSBrown) [inline src/interface/meshblock_data.hpp:242]: 👍🏻

File: src/interface/update.hpp
Comment (JoshuaSBrown) [inline src/interface/update.hpp]: I think that in1 and in2 could be made const, though that would require adding const to the methods in the meshblock_data class and what ever other class is passed in. The PackVariables methods that I have seen are creating a copy and returning it so I don't really see why this can't be done.

Comment (jdolence) [reply src/interface/update.hpp]: It's possible I'm misinterpreting something, but my belief is that the `PackVariables` calls don't work with a `const T *` because the caching modifies the internal state of the object.  So the `PackVariables` calls can't be marked `const`.  I _think_ you could use `T* const in1`, for example, since that just implies the pointer can't be changed, rather than implying that the object pointed to can't be changed.  I'm not sure marking it `T* const` is very helpful, though.

File: tst/unit/test_data_collection.cpp
Comment (JoshuaSBrown) [inline tst/unit/test_data_collection.cpp]: Is this reset necessary? 

Ok, never mind I see, the second variable is changed, but then why are you resetting v1 and v3?

File: src/interface/mesh_data.hpp
Comment (JoshuaSBrown) [inline src/interface/mesh_data.hpp:126]: Shared pointers do not guarantee you you are not pointing to null. It would be safer to make sure that src is valid before calling it's method.

**Review by JoshuaSBrown — CHANGES_REQUESTED** [review]
This is great, an improvement in const correctness and a nice unit test. I would also like to see a unit test for the Sum method, then I will be happy.

**Comment by jdolence** [comment]
> This is great, an improvement in const correctness and a nice unit test. I would also like to see a unit test for the Sum method, then I will be happy.

Thanks.  I'll address the other comments, but I'd push back on the unit test for `SumData`.  The advection tests already make use of `SumData` in several forms and I'd say provide adequate coverage of the method until/unless a problem is encountered that motivates a specific unit test.  A unit test of the method as written would be quite painful to produce, I think, because it's fairly interwoven with other pieces of machinery.  What do you think @JoshuaSBrown?

**Comment by jdolence** [comment]
Just a note to mention that the full pipeline was run and passes.

**Comment by JoshuaSBrown** [comment]
> > This is great, an improvement in const correctness and a nice unit test. I would also like to see a unit test for the Sum method, then I will be happy.
> 
> Thanks. I'll address the other comments, but I'd push back on the unit test for `SumData`. The advection tests already make use of `SumData` in several forms and I'd say provide adequate coverage of the method until/unless a problem is encountered that motivates a specific unit test. A unit test of the method as written would be quite painful to produce, I think, because it's fairly interwoven with other pieces of machinery. What do you think @JoshuaSBrown?

There are advantageous to writing unit tests other than code coverage, it typically helps with the development of class interface. However, in this case I take your point. It's not worth it.

File: src/interface/update.hpp
Comment (JoshuaSBrown) [reply src/interface/update.hpp]: Oh, I see what you mean, that's unfortunate, I skimmed over the helper methods in meshblock_data. Generally I think it is good to use const liberally even if it is only fixing the pointer address, but like you said in this instance the benefits are... minimal.

**Review by JoshuaSBrown — APPROVED** [review]

File: docs/interface/state.md
Comment (Yurlungur) [inline docs/interface/state.md:65]: This is great. A much-needed feature. :+1:

File: src/interface/update.hpp
Comment (Yurlungur) [inline src/interface/update.hpp]: I like this. This is more general and more useful.

Nitpick on the name: This isn't really a sum, it's more like a fuse-multiply-add. How about `TaskStatus FMAData` or something similar?

Also should `w1` and `w2` default to `1.0`?

**Review by Yurlungur — APPROVED** [review]
Thanks for this. This is broadly good cleanup and useful changes.

**Comment by AndrewGaspar** [comment]
> Also changed a lot of the API to T * instead of std::shared_ptr<T> &. const T * would have been better, but the caching we do for packing prevents the usage of const.

GOOD

File: src/interface/update.hpp
Comment (jdolence) [reply src/interface/update.hpp]: It's not really a FMA either, is it?  How about `TaskStatus WeightedSumData`?  I think I like the idea to default the weights to 1...

Comment (jdolence) [reply src/interface/update.hpp]: @Yurlungur pushed a change here.   Look OK?

Comment (Yurlungur) [reply src/interface/update.hpp]: :+1: Yep I like it.

**Comment by jdolence** [comment]
@pgrete, this PR has breaking changes so I want to make sure you see them before this is merged.  Should be a quick review, I think.

File: src/interface/mesh_data.hpp
Comment (pgrete) [reply src/interface/mesh_data.hpp:126]: Wouldn't `if (src.get() == nullptr) {` be more explicit?

File: src/interface/mesh_data.hpp
Comment (pgrete) [inline src/interface/mesh_data.hpp]: this is missing a return statement isn't it?

File: src/interface/meshblock_data.hpp
Comment (pgrete) [inline src/interface/meshblock_data.hpp:114]: What's our current policy with respect to having the same name being used for both a cell variables as well as a SparseVariable?
If that's possible the error handling logic may need an update.

**Review by pgrete — CHANGES_REQUESTED** [review]
Looks good! I only have some minor comments and I think the missing return is actually a bug. that would result in that function always returning true.

File: src/interface/meshblock_data.hpp
Comment (jdolence) [reply src/interface/meshblock_data.hpp:114]: The intent, I believe, is to disallow any two variables from having the same name, sparse or otherwise.

File: src/interface/mesh_data.hpp
Comment (jdolence) [reply src/interface/mesh_data.hpp]: haha.  nice catch.

**Comment by jdolence** [comment]
fixed those issues @pgrete.  thanks for the review (and catching that bug!).  Mind approving so I can merge once the tests pass?

**Comment by jdolence** [comment]
Given the nature of @pgrete's comments and the fact that I've addressed them, I'm just going to pull the trigger.  Thanks for the helpful reviews all!

**Pull request merged**
